File: 06_agents/07_parallel_queries.py
# 07_parallel_queries.py
# Send Parallel Queries to Ollama in Python
# Pairs with 07_parallel_queries.R
# Tim Fraser
#
# Can we send queries to multiple agents at once? Yes!
#
# This script shows how to run basic qualitative coding / content analysis
# very quickly with a local LLM. We'll send many requests in parallel, then
# clean the outputs into standardized labels we can analyze.

# If you haven't already, install these packages...
# pip install requests pandas

# 0. SETUP ###################################

## 0.1 Load Packages #################################

import time  # for timing parallel requests
from concurrent.futures import ThreadPoolExecutor  # for parallel API calls

import pandas as pd  # for reading and sampling feedback data
import requests  # for HTTP requests

## 0.2 Read Data #################################

# Text to classify:
# Let's take a random sample of 10 pieces of feedback.
texts = pd.read_csv("06_agents/07_feedback.csv").sample(n=10, random_state=42)

# 1. CONCEPT ###################################

# Here's an example where we can do basic "qualitative coding"
# (also called content analysis) very quickly.
# Content analysis is the process of examining text and categorizing it
# into meaningful themes or labels.
#
# Sociologists, anthropologists, and market researchers have used this
# approach for decades to analyze interviews, focus groups, and reviews.
#
# You already do this informally in daily life:
# a friend describes a movie, and you mentally tag it as
# "they liked it" or "they didn't like it."
#
# If we provide an LLM with clear, standardized instructions,
# it can help us do this classification at scale.


def get_request(content, prompt, model):
    """Build URL + request body for a local Ollama chat call."""
    port = 11434
    ollama_host = f"http://localhost:{port}"
    url = f"{ollama_host}/api/chat"

    # We use /api/chat so we can send multiple messages
    # (system instructions + user content), not just one prompt.
    body = {
        "model": model,
        "messages": [
            {"role": "system", "content": prompt},
            {"role": "user", "content": content},
        ],
        "stream": False,
    }
    return url, body


def req_perform(content, prompt, model):
    """Send one POST request and return assistant text."""
    url, body = get_request(content=content, prompt=prompt, model=model)
    response = requests.post(url, json=body, timeout=120)
    response.raise_for_status()
    return response.json()["message"]["content"]


# Create a system prompt asking for JSON-style output
# with only the sentiment label.
prompt = (
    "Your only task/role is to evaluate the sentiment of product reviews provided by the user. "
    "Your response should simply be 'positive', 'negative', or 'other'. "
    "Do not include any other text in your response. "
    "Do not include any extra formatting like '```json' in your response. "
    "The JSON string should have the following format: "
    '{"sentiment":"<sentiment of the review>"}'
)

model = "smollm2:1.7b"

# 2. TEST ONE REQUEST ###################################

# Do one quick test request.
test_resp = req_perform(content="booo", prompt=prompt, model=model)
print("Test response:")
print(test_resp)
print()

# 3. SEND REQUESTS IN PARALLEL ##########################

feedback_list = texts["feedback"].astype(str).tolist()

# Send all requests in parallel and time the operation.
start_time = time.time()
with ThreadPoolExecutor(max_workers=min(10, len(feedback_list))) as executor:
    responses = list(executor.map(lambda text: req_perform(text, prompt, model), feedback_list))
elapsed = time.time() - start_time

print(f"Time taken to send {len(feedback_list)} requests: {elapsed:.2f} seconds")

# View the results.
print("Raw responses:")
print(responses)

# 4. CLEAN SENTIMENT LABELS ############################

# Usually there are a few formatting deviations,
# so we use string extraction to get clean labels.
# In regex pattern matching, "|" means OR.
sentiments = (
    pd.Series(responses, name="response")
    .str.lower()
    .str.extract(r"(positive|negative|other)", expand=False)
)
print(sentiments)

File: 06_agents/07_parallel_queries.R
# 07_parallel_queries.R

# Can we send queries to multiple agents at once? Yes!

# Load packages
library(httr2) # for HTTP requests
library(dplyr) # for data wrangling
library(readr) # for reading text files
library(jsonlite) # for JSON handling
library(purrr) # for list-wise operations
library(stringr) # for string extraction

# 1. CONCEPT ###################################

# Here's an example where we can do basic 'qualitative coding' / 'content analysis'very very quickly.
# Content analysis is the process of systematically examining and categorizing textual data to identify patterns, themes, and insights.
# Sociologists, anthropologists, and market researchers have been doing this for decades.
# It's the primary way to analyze focus group, interviews, reviews, and other qualitative data.
# You as a human being have been doing this your whole life, without even knowing it!

# A friend tells you what they thought of a movie. You categorize in your head as "they liked it" or "they didn't like it".
# If we provide an LLM with really good, standardized instructions,
# it can do this for us very quickly.

# text to classify
texts = read_csv("06_agents/07_feedback.csv") %>%
   sample_n(10) # Let's take a random sample of 10

# Make a quick wrapper function for a request
get_request = function(CONTENT, PROMPT, MODEL){
    
    PORT = 11434
    OLLAMA_HOST = paste0("http://localhost:", PORT)
    url = paste0("http://localhost:", PORT, "/api/chat")

    # We're going to use the /api/chat endpoint,
    # which allows us to send a list of messages, not just a single prompt.

    # Construct the request body as a list
    body = list(
        model = MODEL, # Model name
        messages = list(
            list(role = "system", content = PROMPT),
            list(role = "user", content = CONTENT)  ),
        stream = FALSE       # Non-streaming response
    )

    # Build and send the POST request to the REST API
    req = httr2::request(url) %>%
        req_body_json(body) %>%   # Attach the JSON body
        req_method("POST")   # Set HTTP method

    # Return the request object, BEFORE it is sent.
    return(req)
}


# Create a system prompt, asking it to return a json string
PROMPT = paste0(
    "Your only task/role is to evaluate the sentiment of product reviews provided by the user. ",
    "Your response should simply be 'positive', 'negative', or 'other'. ",
    "Return a JSON string of your response. ",
    "Do not include any other text in your response. ",
    "Do not include any extra formatting like  '```json' in your response. ",
    "The JSON string should have the following format: ",
    "{",
    "  \"sentiment\": \"<sentiment of the review>\",",
    "}"
)

MODEL = "smollm2:1.7b"

# Do one quick test request
resp = get_request(CONTENT = "booo", PROMPT = PROMPT, MODEL = MODEL) |>
    req_perform()
    
# Check the output!
resp |> resp_body_json() |> with(message) |> with(content)


# Create a list of request objects, BEFORE they are sent.
reqs = texts$feedback |>
   map(~get_request(CONTENT = ., PROMPT = PROMPT, MODEL = MODEL)) 

# Send the requests in parallel and time the operation
time = system.time({
    # make parallel requests and get response
    resps = req_perform_parallel(reqs)
})
cat("\n Time taken to send", length(reqs), "requests: ", round(time[3], 2), " seconds\n")

# View the results!
resps |> 
   map(~resp_body_json(.x) |> with(message) |> with(content) ) 

# Usually, there will be a few deviations in formatting,
# so you'll probably want to use string extraction
# to get a nice clean result.

resps |> 
   # For each response, extract the content
   map_chr(~resp_body_json(.x) |> with(message) |> with(content) )  %>%
   # Standardize to lower case, just in case
   tolower() %>%
   # Extract the keywords you asked them to code
   # In REGEX string pattern matching,
   # "|" means OR
   stringr::str_extract("positive|negative|other")

rm(list = ls())